### DISCUSSION TITLE
Kotlin Multiplatform support

### DISCUSSION THREAD

**Discussion by kizitonwose** [kizitonwose/Calendar#554] [Announcements]
Edit: [Kotlin multiplatform calendar library alpha release is now available.](https://github.com/kizitonwose/Calendar/discussions/554#discussioncomment-9975625)

---

Over the weekend, I had some time to investigate what is needed to add multiplatform support to the library, seeing that some users have requested it over the past few months. I was able to get a rough example of the existing compose `Example1Page` up and running. Screenshots from the desktop app and browser are shown below, next to the `Example1Page` from the Android sample app. This iOS sample app also works, but I moved a bunch of source files around which broke the compilation and I did not want to spend time fixing it just for the demo.

|Android|Desktop|Browser|
|:-:|:-:|:-:|
|![Android](https://github.com/kizitonwose/Calendar/assets/15170090/34a36a13-b7dc-414e-adc7-d7dc0d46920e)|![Desktop](https://github.com/kizitonwose/Calendar/assets/15170090/d97b08a2-38f7-45fe-80c0-0fbe794755bb)|![Browser](https://github.com/kizitonwose/Calendar/assets/15170090/c70ea490-e695-4e89-b632-df8094a2d2cf)|

I have some thoughts on how to approach this, please feel free to chime in.

The library presently has support for the view system, and that is not going away. The current compose artifact is also not going away either as together with the view system library, they use the `java.time` APIs and I intend to keep them that way. The `kotlinx-datetime` library has come a long way, but it is not yet at the same level as the `java.time` APIs, so I do not want to force the library users to use the `kotlinx-datetime` library, especially on pure Android projects without multiplatform setup.

I am considering publishing the multiplatform library which will use the `kotlinx-datetime` library under a different module name, like `com.kizitonwose.calendar:compose-multiplatform`, so the Android `view` and `compose` libraries using the `java.time` APIs will keep existing in their modules `com.kizitonwose.calendar:view` and `com.kizitonwose.calendar:compose` respectively, but this means duplicating a lot of compose code, so I am still thinking about the right approach and open to suggestions.

Regarding the timeline for when this could be out, at least in alpha, I can't give anything specific right now, but given my current schedule, maybe in 2-ish months. 

**I kindly ask you to refrain from posting comments solely asking about the release date, as such comments do not add any value.**

**Comment by abdallahmehiz** [comment]
Why not create a module that uses both `java.time` and `kotlinx.datetime`? one for the jvm and the other for native?

**Reply by kizitonwose** [reply]
@abdallahmehiz can you elaborate further? Do you mean to bundle both `java.time` and `kotlinx.datetime` logic in one module, then expose different composables?

**Reply by abdallahmehiz** [reply]
Lol sorry i think i was half asleep when i wrote this yesterday...
I thing what i was trying to say is just create a common API, maybe an interface that has all of your needed functionality for handling dates then implement them for both native and jvm with the expect actual mechanism.
You're probably already doing that though...

**Reply by kizitonwose** [reply]
Went a bit in this direction, but it started to get quite messy, might try [the approach with generics](https://github.com/kizitonwose/Calendar/discussions/554#discussioncomment-9874821).

**Comment by kalist28** [comment]
I think it’s worth switching completely to ``` kotlinx.datetime ```, and leaving functions from ``` Java.datetime ``` as duplicate functions from the multiplatform but with mapping from kotlinx.datetime

Kotlinx.datetime has a convenient translation API to Java.datetime and back to the jvm target.

You will get two options for Android calendar:compose:0.0.0 with Java.datetime and calendar:compose-multiplatform:0.0.0 with kotlinx.datetime

**Reply by kizitonwose** [reply]
Yep, this could work, I will have to make a lot of the core classes generic to hold the `kotlinx.datetime` type or the `Java.datetime` type depending on the library type.

**Reply by bmc08gt** [reply]
Sorry for the massive hiatus. Yes that is what I did (Kotlinx date time).  The code for it was in my PR here for my own project - https://github.com/bmcreations/airbnb-multiplatform-sample/pull/3/files#diff-697f70cdd88ba88fe77eebda60c7e143f6ad1286bca75017421e93ad84fb87df

**Comment by kizitonwose** [comment]
Published the first alpha for the multiplatform artifact today, it has the same API as the existing compose calendar, so the existing compose documentation should be sufficient to get started. 

Note that this is based on compose UI 1.7.0-beta03.

```kotlin
commonMain.dependencies {
    implementation("com.kizitonwose.calendar:compose-multiplatform:2.6.0-alpha02")
}
```

Examples can be viewed online at: https://calendar.kizitonwose.dev, the source code is available in the [multiplatform](https://github.com/kizitonwose/Calendar/tree/multiplatform) branch.

**Comment by kizitonwose** [comment]
The multiplatform library code is now merged into the main branch in #561, this discussion will therefore be closed.
